File: tests/ten_runtime/integration/python/async_get_set_prop_2_python/async_get_set_prop_2_python_app/ten_packages/extension/default_async_extension_python/extension.py
#
# Copyright © 2025 Agora
# This file is part of TEN Framework, an open source project.
# Licensed under the Apache License, Version 2.0, with certain conditions.
# Refer to the "LICENSE" file in the root directory for more information.
#
import json
from ten_runtime import AsyncExtension, AsyncTenEnv, Cmd, CmdResult


class DefaultAsyncExtension(AsyncExtension):

    async def on_init(self, ten_env: AsyncTenEnv) -> None:
        test_prop_json_str, _ = await ten_env.get_property_to_json(
            "params.billing_type"
        )
        test_prop_json = json.loads(test_prop_json_str)
        assert test_prop_json == "agent_billing"

        test_prop_json_str, _ = await ten_env.get_property_to_json()
        test_prop_json = json.loads(test_prop_json_str)
        assert test_prop_json["params"]["billing_type"] == "agent_billing"

        test_prop_json_str, _ = await ten_env.get_property_to_json("")
        test_prop_json = json.loads(test_prop_json_str)
        assert test_prop_json["params"]["billing_type"] == "agent_billing"

    async def on_cmd(self, ten_env: AsyncTenEnv, cmd: Cmd) -> None:
        # Send a new command to other extensions and wait for the result. The
        # result will be returned to the original sender.
        new_cmd = Cmd.create("hello")
        cmd_result, _ = await ten_env.send_cmd(new_cmd)
        assert cmd_result is not None

        cmd_result_json, _ = cmd_result.get_property_to_json()

        new_result = CmdResult.create(cmd_result.get_status_code(), cmd)
        new_result.set_property_from_json(None, cmd_result_json)

        await ten_env.return_result(new_result)

File: tests/ten_runtime/integration/go/no_dest_go/no_dest_go_app/ten_packages/extension/extension_a/extension.go
//
// Copyright © 2025 Agora
// This file is part of TEN Framework, an open source project.
// Licensed under the Apache License, Version 2.0, with certain conditions.
// Refer to the "LICENSE" file in the root directory for more information.
//

package default_extension_go

import (
	ten "ten_framework/ten_runtime"
)

type baseExtension struct {
	ten.DefaultExtension
}

type aExtension struct {
	baseExtension

	counter int
}

func newAExtension(name string) ten.Extension {
	return &aExtension{counter: 0}
}

func (ext *aExtension) OnStart(tenEnv ten.TenEnv) {
	tenEnv.Log(ten.LogLevelDebug, "OnStart")

	noDestCmd, _ := ten.NewCmd("unknownCmd")
	tenEnv.SendCmd(noDestCmd, func(te ten.TenEnv, cr ten.CmdResult, err error) {
		if err == nil {
			panic("SendCmd should fail if no destination is found.")
		}

		tenEnv.Log(ten.LogLevelInfo, "SendCmd failed as expected, err: " + err.Error())

		ext.counter++

		if ext.counter == 4 {
			tenEnv.OnStartDone()
		}
	})

	noDestData, _ := ten.NewData("unknownData")
	tenEnv.SendData(noDestData, func(te ten.TenEnv, err error) {
		if err == nil {
			panic("SendData should fail if no destination is found.")
		}

		tenEnv.Log(ten.LogLevelInfo, "SendData failed as expected, err: " + err.Error())

		ext.counter++

		if ext.counter == 4 {
			tenEnv.OnStartDone()
		}
	})

	noDestVideoFrame, _ := ten.NewVideoFrame("unknownVideoFrame")
	tenEnv.SendVideoFrame(noDestVideoFrame, func(te ten.TenEnv, err error) {
		if err == nil {
			panic("SendVideoFrame should fail if no destination is found.")
		}

		tenEnv.Log(ten.LogLevelInfo, "SendVideoFrame failed as expected, err: " + err.Error())

		ext.counter++

		if ext.counter == 4 {
			tenEnv.OnStartDone()
		}
	})

	noDestAudioFrame, _ := ten.NewAudioFrame("unknownAudioFrame")
	tenEnv.SendAudioFrame(noDestAudioFrame, func(te ten.TenEnv, err error) {
		if err == nil {
			panic("SendAudioFrame should fail if no destination is found.")
		}

		tenEnv.Log(ten.LogLevelInfo, "SendAudioFrame failed as expected, err: " + err.Error())

		ext.counter++

		if ext.counter == 4 {
			tenEnv.OnStartDone()
		}
	})
}

func (ext *aExtension) OnStop(tenEnv ten.TenEnv) {
	tenEnv.Log(ten.LogLevelDebug, "OnStop")

	tenEnv.OnStopDone()
}

func (p *aExtension) OnCmd(
	tenEnv ten.TenEnv,
	cmd ten.Cmd,
) {
	cmdResult, _ := ten.NewCmdResult(ten.StatusCodeOk, cmd)
	cmdResult.SetPropertyString("detail", "okok")
	err := tenEnv.ReturnResult(cmdResult, nil)
	if err != nil {
		panic("ReturnResult failed")
	}
}

func init() {
	// Register addon
	err := ten.RegisterAddonAsExtension(
		"extension_a",
		ten.NewDefaultExtensionAddon(newAExtension),
	)
	if err != nil {
		panic("Failed to register addon.")
	}
}
